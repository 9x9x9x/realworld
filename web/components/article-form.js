import React from 'react';
import PropTypes from 'prop-types';
import { Formik, Form, ErrorMessage, Field } from 'formik';
import { FormikStatusErrors, FormikSubmitButton } from './formik';
import { TagsInput } from '../containers';

export function ArticleForm(props) {
  return (
    <Formik
      validationSchema={props.validationSchema}
      enableReinitialize
      initialValues={props.initialValues}
      onSubmit={props.onSubmit}
    >
      <Form>
        <ul className="error-messages">
          <ErrorMessage component="li" name="slug" />
          <ErrorMessage component="li" name="input.title" />
          <ErrorMessage component="li" name="input.description" />
          <ErrorMessage component="li" name="input.body" />
          <ErrorMessage component="li" name="input.tagId" />
          <FormikStatusErrors />
        </ul>

        <fieldset>
          <fieldset className="form-group">
            <label>Article Title</label>
            <Field
              name="input.title"
              type="text"
              className="form-control form-control-lg"
              placeholder="How to build webapps that scale"
            />
          </fieldset>
          <fieldset className="form-group">
            <label>Description</label>
            <Field
              name="input.description"
              type="text"
              className="form-control"
              placeholder="Rock solid process you can follow when building your apps"
            />
          </fieldset>
          <fieldset className="form-group">
            <label>Body</label>
            <Field
              name="input.body"
              as="textarea"
              className="form-control"
              rows={8}
              placeholder={`# Introducing RealWorld.\n\nIt's a great solution for learning how other frameworks work.`}
            />
          </fieldset>
          <fieldset className="form-group">
            <label>Tags</label>
            <TagsInput name="input.tagIds" />
          </fieldset>
          <FormikSubmitButton
            disabled={props.disabled}
            className="btn btn-lg pull-xs-right btn-primary"
          >
            Publish Article
          </FormikSubmitButton>
        </fieldset>
      </Form>
    </Formik>
  );
}

ArticleForm.propTypes = {
  disabled: PropTypes.bool,
  validationSchema: PropTypes.object.isRequired,
  initialValues: PropTypes.object.isRequired,
  onSubmit: PropTypes.func.isRequired
};
