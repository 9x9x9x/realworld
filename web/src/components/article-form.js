import React from 'react';
import PropTypes from 'prop-types';
import { Formik, Form, ErrorMessage, Field } from 'formik';
import { FormikStatusErrors } from './formik/formik-status-errors';
import { FormikSubmitButton } from './formik/formik-submit-button';
import { TagsInput } from '../containers/tags-input';

export function ArticleForm(props) {
  return (
    <Formik
      validationSchema={props.validationSchema}
      enableReinitialize
      initialValues={props.initialValues}
      onSubmit={props.onSubmit}
    >
      <Form id="article-form">
        <ul className="error-messages">
          <ErrorMessage component="li" name="slug" />
          <ErrorMessage component="li" name="input.title" />
          <ErrorMessage component="li" name="input.description" />
          <ErrorMessage component="li" name="input.body" />
          <ErrorMessage component="li" name="input.tagIds" />
          <FormikStatusErrors />
        </ul>

        <fieldset>
          <fieldset className="form-group">
            <label htmlFor="article-form-title-input">Title</label>
            <Field
              name="input.title"
              type="text"
              id="article-form-title-input"
              className="form-control form-control-lg"
              placeholder="How to build webapps that scale"
            />
          </fieldset>
          <fieldset className="form-group">
            <label htmlFor="article-form-description-input">Description</label>
            <Field
              name="input.description"
              type="text"
              id="article-form-description-input"
              className="form-control"
              placeholder="Web development technologies have evolved at an incredible clip over the past few years."
            />
          </fieldset>
          <fieldset className="form-group">
            <label htmlFor="article-form-body-textarea">Body</label>
            <Field
              name="input.body"
              as="textarea"
              className="form-control"
              id="article-form-body-textarea"
              rows={8}
              placeholder={`# Introducing RealWorld.\n\nIt's a great solution for learning how other frameworks work.`}
            />
          </fieldset>
          <fieldset className="form-group">
            <label htmlFor="article-form-tags-ids-input">Tags</label>
            <TagsInput name="input.tagIds" id="article-form-tags-ids-input" />
          </fieldset>
          <FormikSubmitButton
            disabled={props.disabled}
            className="btn btn-lg pull-xs-right btn-primary"
          >
            Publish Article
          </FormikSubmitButton>
        </fieldset>
      </Form>
    </Formik>
  );
}

ArticleForm.propTypes = {
  disabled: PropTypes.bool,
  validationSchema: PropTypes.object.isRequired,
  initialValues: PropTypes.object.isRequired,
  onSubmit: PropTypes.func.isRequired
};
