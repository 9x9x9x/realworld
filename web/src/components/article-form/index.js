import React from 'react';
import PropTypes from 'prop-types';
import { Formik, Form, ErrorMessage, Field } from 'formik';
import { FormikStatusErrors } from '../formik-status-errors';
import { FormikSubmitButton } from '../formik-submit-button';
import { TagsInput } from '../../containers/tags-input';
import * as Yup from 'yup';

const createValidationSchema = Yup.object({
  input: Yup.object({
    title: Yup.string()
      .label('Title')
      .required(),
    description: Yup.string()
      .label('Description')
      .required(),
    body: Yup.string()
      .label('Body')
      .required(),
    tagIds: Yup.array(Yup.string())
      .label('Tags')
      .test('', '${path} is a required field', value => Array.isArray(value))
  })
});

const updateValidationSchema = Yup.object({
  slug: Yup.string()
    .label('Slug')
    .required(),
  input: Yup.object({
    title: Yup.string()
      .label('Title')
      .required(),
    description: Yup.string()
      .label('Description')
      .required(),
    body: Yup.string()
      .label('Body')
      .required(),
    tagIds: Yup.array(Yup.string())
      .label('Tags')
      .test('', '${path} is a required field', value => Array.isArray(value))
  })
});

export function ArticleForm({ initialValues, onSubmit, disabled }) {
  const validationSchema = Object.prototype.hasOwnProperty.call(
    initialValues,
    'slug'
  )
    ? updateValidationSchema
    : createValidationSchema;
  return (
    <div className="container page">
      <div className="row">
        <div className="col-md-10 offset-md-1 col-xs-12">
          <Formik
            validationSchema={validationSchema}
            enableReinitialize
            initialValues={initialValues}
            onSubmit={onSubmit}
          >
            <Form id="article-form">
              <ul className="error-messages">
                <ErrorMessage component="li" name="slug" />
                <ErrorMessage component="li" name="input.title" />
                <ErrorMessage component="li" name="input.description" />
                <ErrorMessage component="li" name="input.body" />
                <ErrorMessage component="li" name="input.tagIds" />
                <FormikStatusErrors />
              </ul>

              <fieldset>
                <fieldset className="form-group">
                  <label htmlFor="article-form-title-input">Title</label>
                  <Field
                    name="input.title"
                    type="text"
                    id="article-form-title-input"
                    className="form-control form-control-lg"
                    placeholder="How to build webapps that scale"
                  />
                </fieldset>
                <fieldset className="form-group">
                  <label htmlFor="article-form-description-input">
                    Description
                  </label>
                  <Field
                    name="input.description"
                    type="text"
                    id="article-form-description-input"
                    className="form-control"
                    placeholder="Web development technologies have evolved at an incredible clip over the past few years."
                  />
                </fieldset>
                <fieldset className="form-group">
                  <label htmlFor="article-form-body-textarea">Body</label>
                  <Field
                    name="input.body"
                    as="textarea"
                    className="form-control"
                    id="article-form-body-textarea"
                    rows={8}
                    placeholder={`# Introducing RealWorld.\n\nIt's a great solution for learning how other frameworks work.`}
                  />
                </fieldset>
                <fieldset className="form-group">
                  <label htmlFor="article-form-tags-ids-input">Tags</label>
                  <TagsInput
                    name="input.tagIds"
                    id="article-form-tags-ids-input"
                  />
                </fieldset>
                <FormikSubmitButton
                  disabled={disabled}
                  className="btn btn-lg pull-xs-right btn-primary"
                >
                  Publish Article
                </FormikSubmitButton>
              </fieldset>
            </Form>
          </Formik>
        </div>
      </div>
    </div>
  );
}

ArticleForm.propTypes = {
  disabled: PropTypes.bool,
  initialValues: PropTypes.object.isRequired,
  onSubmit: PropTypes.func.isRequired
};
