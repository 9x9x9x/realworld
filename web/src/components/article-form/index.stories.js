import React from 'react';
import { ArticleForm } from '.';
import { action } from '@storybook/addon-actions';
import { withApolloClient } from 'storybook-addon-apollo-client';

export default {
  title: 'Forms/ArticleForm',
  component: ArticleForm,
  decorators: [withApolloClient],
};

export const renders = () => <ArticleForm onSubmit={action('onSubmit')} />;
