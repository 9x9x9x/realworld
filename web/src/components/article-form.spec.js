import React from 'react';
import { render, fireEvent, wait } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import * as Yup from 'yup';
import { MockedProvider } from '@apollo/react-testing';

import { ArticleForm } from './article-form';

const validationSchema = Yup.object({
  input: Yup.object({
    title: Yup.string()
      .label('Title')
      .required(),
    description: Yup.string()
      .label('Description')
      .required(),
    body: Yup.string()
      .label('Body')
      .required(),
    tagIds: Yup.array(Yup.string())
      .label('Tags')
      .test('', '${path} is a required field', value => Array.isArray(value))
  })
});

describe('ArticleForm', () => {
  function setup({
    handleSubmit = () => undefined,
    initialValues = {
      input: { title: '', description: '', body: '', tagIds: [] }
    }
  } = {}) {
    return render(
      <MockedProvider>
        <ArticleForm
          validationSchema={validationSchema}
          initialValues={initialValues}
          onSubmit={handleSubmit}
        />
      </MockedProvider>
    );
  }

  it('renders', async () => {
    const { asFragment } = setup();
    await wait(() => {
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it('displays errors on submit', async () => {
    const { getByRole, getAllByText } = setup();

    await wait(() => {
      const submitButton = getByRole('button');
      fireEvent.click(submitButton);
      const errors = getAllByText(/is a required field$/);

      expect(errors).toHaveLength(3);
    });
  });

  it('calls onSubmit when valid on submit', async () => {
    let handleSubmit = jest.fn();

    const { getByRole, getByLabelText } = setup({
      handleSubmit
    });

    await wait(() => {
      fireEvent.change(getByLabelText(/title/i), {
        target: { value: 'How to build webapps that scale' }
      });
    });

    await wait(() => {
      fireEvent.change(getByLabelText(/description/i), {
        target: {
          value:
            'Web development technologies have evolved at an incredible clip over the past few years.'
        }
      });
    });

    await wait(() => {
      fireEvent.change(getByLabelText(/body/i), {
        target: {
          value:
            "## Introducing RealWorld.\nIt's a great solution for learning how other frameworks work."
        }
      });
    });

    await wait(() => {
      const submitButton = getByRole('button');

      fireEvent.click(submitButton);

      expect(handleSubmit).toHaveBeenCalled();
    });
  });
});
