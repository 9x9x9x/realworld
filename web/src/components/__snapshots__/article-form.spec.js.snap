// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleForm renders 1`] = `
<DocumentFragment>
  <form
    action="#"
    id="article-form"
  >
    <ul
      class="error-messages"
    />
    <fieldset>
      <fieldset
        class="form-group"
      >
        <label
          for="article-form-title-input"
        >
          Title
        </label>
        <input
          class="form-control form-control-lg"
          id="article-form-title-input"
          name="input.title"
          placeholder="How to build webapps that scale"
          type="text"
          value=""
        />
      </fieldset>
      <fieldset
        class="form-group"
      >
        <label
          for="article-form-description-input"
        >
          Description
        </label>
        <input
          class="form-control"
          id="article-form-description-input"
          name="input.description"
          placeholder="Rock solid process you can follow when building your apps"
          type="text"
          value=""
        />
      </fieldset>
      <fieldset
        class="form-group"
      >
        <label
          for="article-form-body-textarea"
        >
          Body
        </label>
        <textarea
          class="form-control"
          id="article-form-body-textarea"
          name="input.body"
          placeholder="# Introducing RealWorld.

It's a great solution for learning how other frameworks work."
          rows="8"
        />
      </fieldset>
      <fieldset
        class="form-group"
      >
        <label
          for="article-form-tags-ids-input"
        >
          Tags
        </label>
        <p>
          <input
            class="form-control"
            id="article-form-tags-ids-input"
            list="tags"
            name="input.tagIds"
            placeholder="Press enter to add tag to list"
            type="text"
            value=""
          />
          <datalist
            id="tags"
          />
        </p>
        <div
          class="tag-list"
        />
      </fieldset>
      <button
        class="btn btn-lg pull-xs-right btn-primary"
        type="submit"
      >
        Publish Article
      </button>
    </fieldset>
  </form>
</DocumentFragment>
`;
